rpcs3/Emu/Cell/Modules/sys_libc_.cpp

```diff
@@ -129,82 +189,160 @@ vm::ptr<void> _sys_memmove(vm::ptr<void> dst, vm::cptr<void> src, u32 size)
 	return dst;
 }
 
-s64 _sys_strlen(vm::cptr<char> str)
+u32 _sys_strlen(vm::cptr<char> str)
 {
 	sysPrxForUser.trace("_sys_strlen(str=%s)", str);
 
-	return std::strlen(str.get_ptr());
+	if (!str)
+	{
+		return 0;
+	}
+
+	for (u32 i = 0;; i++)
+	{
+		if (str[i] == '\0')
+		{
+			return i;
+		}
+	}
 }
 
 s32 _sys_strcmp(vm::cptr<char> str1, vm::cptr<char> str2)
 {
 	sysPrxForUser.trace("_sys_strcmp(str1=%s, str2=%s)", str1, str2);
 
-	return std::strcmp(str1.get_ptr(), str2.get_ptr());
+	for (u32 i = 0;; i++)
```

Comment:
I'm trying to understand a bit of RPCS3 code. This change seems odd. What's the actual difference between using the ones from the standard system libraries and reimplementing? Are the system libraries inefficient?

Reply:
I don't understand why, but it's a bad idea to pass virtual PS3 memory to system calls.

Reply:
Standard doesn't specify exact return values (-1 or 1 for example) and defines many undefined behavior conditions.

Reply:
I see, now it makes more sense. Thanks.